File: c/binary-search/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length)
{
  if (length == 0) return NULL;
  if (value < arr[0] || value > arr[length-1]) return NULL;

  size_t low = 0;
  size_t high = length - 1;
  while(low <= high) {
    size_t mid = low + (high - low) / 2;
    if (arr[mid] == value) {
      return &arr[mid];
    } else if (arr[mid] < value) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return NULL;
}

File: zig/binary-search/binary_search.zig
// Take a look at the tests, you might have to change the function arguments

pub fn binarySearch(comptime T: type, target: T, items: []const T) ?usize {
    if (items.len == 0) return null;
    if (target < items[0] or target > items[items.len - 1]) return null;

    var low: usize = 0;
    var high: usize = items.len - 1;
    while (low <= high) {
        const mid = low + (high - low) / 2;
        if (items[mid] == target) {
            return mid;
        } else if (items[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return null;
}
